tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/AugmentedTaggedMetricRegistry.java

```diff
@@ -0,0 +1,202 @@
+/*
+ * (c) Copyright 2020 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.tritium.metrics.registry;
+
+import com.codahale.metrics.Counter;
+import com.codahale.metrics.Gauge;
+import com.codahale.metrics.Histogram;
+import com.codahale.metrics.Meter;
+import com.codahale.metrics.Metric;
+import com.codahale.metrics.Timer;
+import com.palantir.logsafe.Preconditions;
+import com.palantir.logsafe.Safe;
+import com.palantir.logsafe.SafeArg;
+import com.palantir.logsafe.exceptions.SafeIllegalArgumentException;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.BiConsumer;
+import java.util.function.Supplier;
+
+/**
+ * Any metric created on this registry will be saved into the delegate {@link TaggedMetricRegistry} with an extra tag
+ * added. This ensures any reads from the underlying delegate will have the desired 'augmented' tag.
+ */
+public final class AugmentedTaggedMetricRegistry implements TaggedMetricRegistry {
```

Comment:
what other purposes do you imagine for an `AugmentedTaggedMetricRegistry` besides adding library name/version? I wonder if we could just simplify to having a single   `ProvenanceTaggedMetricRegistry` (its a working title) which takes a name, version and a delegate. This would allow us to do away with a a bunch of boilerplate in the implementation and could make passing registries across libraries safer since you easily access and re-wrap the delegate of another `ProvenanceTaggedMetricRegistry`.